ca_tools/scripts/draw_square.py

```diff
@@ -0,0 +1,209 @@
+
+    def _rotate(self):
+        rospy.loginfo("rotate")
+        self._twist.angular.z = self.ANGULAR_VEL
+        self._twist.linear.x = 0
+
+        while(self._diff_angle > self.ANGLE_THRESHOLD):
+            self._publish()
+            self._update_current_pose()
+            self._get_pose_diff()
+        self._stop()
+
+    def _stop(self):
+        rospy.loginfo("stop")
+        self._twist.angular.z = 0
+        self._twist.linear.x  = 0
+        self._publish()
+
+    def _get_pose_diff(self):
+        q = self._current_pose.orientation
+        _, _, yaw = efq((q.x,q.y,q.z,q.w))
+        self._current_angle = self._wrap_angle(yaw)
+
+        diff_x = (self._goal_pose.position.x - self._current_pose.position.x)
+        diff_y = (self._goal_pose.position.y - self._current_pose.position.y)
+
+        self._goal_angle = self._wrap_angle(math.atan2(diff_y,diff_x))
+        self._diff_angle = self._wrap_angle(self._goal_angle - self._current_angle)
+        self._length_diff = math.sqrt(diff_x*diff_x + diff_y*diff_y)
+
+    def _set_square_goal(self):
+            q = self._current_pose.orientation
+            euler = efq((q.x,q.y,q.z,q.w))
+            euler_yaw = self._wrap_angle(euler[2] + math.pi/2.0)
+            q = qfe(euler[0],euler[1],euler_yaw) 
+            self._goal_pose.position.x = self._current_pose.position.x + self._square_length*math.cos(euler_yaw)
+            self._goal_pose.position.y = self._current_pose.position.y + self._square_length*math.sin(euler_yaw)
+            self._update_controller_goal()
+            return 
+
+    def _get_next_state(self):
+        rospy.loginfo("Get next state")
+        if(self._diff_angle > self.ANGLE_THRESHOLD):
+            self._state.SetState("ROTATE")
+        elif(self._length_diff > self.LENGTH_THRESHOLD):
+            self._state.SetState("MOVE_FORWARD")
+        else:
+            self._state.SetState("STOP")
+            self._currentGoalReached = True
+
+    def _move_forward(self):
+        rospy.loginfo("MF")
+        self._controller_on()
+        t0 = rospy.get_time()
+        distance_travelled = 0
+        self._twist.linear.x = self.LINEAR_VEL
+        self._twist.angular.z = 0
+        while(abs(distance_travelled - self._length_diff) > self.LENGTH_THRESHOLD):
+            self._publish()
+            t1 = rospy.get_time()
+            distance_travelled = (self._twist.linear.x)*(t1-t0)
+        self._length_diff -= distance_travelled
+        self._controller_off()
+        self._stop()
+
+
+    def _update_current_pose(self):
+        self._current_pose = self._ground_truth.get_pose() #Could be better implemented using callbacks
+
+    def _goal_reached(self):
+        rospy.loginfo("goal_reached")
+        if(self._length_diff < self.GOAL_DIST_THRESHOLD ):
+            self._currentGoalReached = True
+            return True
+        else:
+            return False
+
+    
+    _actions = {'MOVE_FORWARD':_move_forward ,'STOP':_stop ,'ROTATE':_rotate}
```

Comment:
what is this used for? where are the variables defined?

Reply:
Made a comment in the code for clarification.